### PR TITLE
Fix concurrency issue in equinox caching

### PR DESCRIPTION
**Pull request by xpomul**
This PR fixes issue #233 

I had a short discussion with @martinlippert a few weeks ago about how to fix this.

From that, I have tried to first implement a lookup map for classes that are currently in the writer queue. This solves the issue of weaving the same class multiple times, because even if the class is not yet written to disk, we have a cache hit and can already reuse that class in the writer queue.

This implementation made things a bit better, but still resulted in the same symptom; it only became rarer.

The problem is most likely that there is no synchronization between cache reads and writes, and it can be the case that the reader has a cache miss, and immediately at that point (before the reader checks the file system), the writer is given the same class to store and it directly starts to store it, which still can result in an incomplete file being present when the reader now checks the file system.

So, I implemented a synchronization mechanism in addition to the queue lookup. This synchronization uses one RWLock per class name, so that we still have optimal throughput as long as no conflicting access occurs.

With this implementation, the RCPTT tests are now rock-solid.

### CODE REVIEW

File: bundles/org.eclipse.equinox.weaving.caching/src/org/eclipse/equinox/weaving/internal/caching/IBundleConstants.java [hunk 1]
```diff
@@ -1,28 +1,34 @@
...
  * @author Heiko Seeberger
  */
 public interface IBundleConstants {
 
-    /**
-     * The symbolic name for this bundle: "org.aspectj.osgi.service.caching".
-     */
-    public static final String BUNDLE_SYMBOLIC_NAME = "org.eclipse.equinox.weaving.caching"; //$NON-NLS-1$
+	/**
+	 * The symbolic name for this bundle: "org.aspectj.osgi.service.caching".
```
[the remark this answers is not in the document]
Comment (xpomul) [reply bundles/org.eclipse.equinox.weaving.caching/src/org/eclipse/equinox/weaving/internal/caching/IBundleConstants.java:26]: ah, that code seems to be quite old and the name seems to be even older :-D
The code format was even not up to date which is why I splitted the PR into two commits to get the format correct at first, but I haven’t noticed this obvious JavaDoc error. I’ll fix it together with any other feedback in the next days...

**Comment by martinlippert** [comment]
What happens if you first add the entry to the lookup map and then add the entry to the cache writer queue (on write)? https://github.com/xpomul/equinox/blob/f4640527b063c75aeca9aa6d6c58ee6f90ab9a32/bundles/org.eclipse.equinox.weaving.caching/src/org/eclipse/equinox/weaving/internal/caching/BundleCachingService.java#L261

**Comment by xpomul** [comment]
> What happens if you first add the entry to the lookup map and then add the entry to the cache writer queue (on write)? https://github.com/xpomul/equinox/blob/f4640527b063c75aeca9aa6d6c58ee6f90ab9a32/bundles/org.eclipse.equinox.weaving.caching/src/org/eclipse/equinox/weaving/internal/caching/BundleCachingService.java#L261

Did the modification locally and ran another round of CI tests. The Truncated Class error still occurs even then
Also, if we’d change the order here, we’d need to take care of the case when the queue is full and does not accept the item. This is why I kept the original order. Because if we add the item to the lookup, but it does not get added to the queue, it will remain in the lookup indefinitely and never be removed.

Now that I think of it, I could add it to the lookup beforehand and remove it again if the queue does not accept it. That should do no harm and might give a few additional lookup cache hits in concurrency situations... (but as said above, it does not fully solve the basic concurrency issues)

**Comment by martinlippert** [comment]
+1 from my side.

**Comment by xpomul** [comment]
> We need a version bump done in org.eclipse.equinox.weaving.caching/META-INF/MANIFEST.MF for the `Bundle-Version`. Update that to `1.2.300.qualifier`

Thanks. Done and merged PR to current state.

**Comment by xpomul** [comment]
(sorry, didn’t want to close :-P)

**Review by tjwatson — APPROVED** [review]

**Pull request merged**